core/src/main/java/org/elasticsearch/search/aggregations/bucket/adjacency/AdjacencyFiltersAggregator.java

```diff
@@ -0,0 +1,226 @@
+
+        public KeyedFilter(String key, QueryBuilder filter) {
+            if (key == null) {
+                throw new IllegalArgumentException("[key] must not be null");
+            }
+            if (filter == null) {
+                throw new IllegalArgumentException("[filter] must not be null");
+            }
+            this.key = key;
+            this.filter = filter;
+        }
+
+        /**
+         * Read from a stream.
+         */
+        public KeyedFilter(StreamInput in) throws IOException {
+            key = in.readString();
+            filter = in.readNamedWriteable(QueryBuilder.class);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(key);
+            out.writeNamedWriteable(filter);
+        }
+
+        public String key() {
+            return key;
+        }
+
+        public QueryBuilder filter() {
+            return filter;
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.field(key, filter);
+            return builder;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(key, filter);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == null) {
+                return false;
+            }
+            if (getClass() != obj.getClass()) {
+                return false;
+            }
+            KeyedFilter other = (KeyedFilter) obj;
+            return Objects.equals(key, other.key) && Objects.equals(filter, other.filter);
+        }
+    }
+
+    private final String[] keys;
+    private Weight[] filters;
+    private final int totalNumKeys;
+    private final int totalNumIntersections;
+    private final String separator;
+
+    public AdjacencyFiltersAggregator(String name, AggregatorFactories factories, String separator, String[] keys, Weight[] filters,
+            SearchContext context, Aggregator parent, List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData)
+            throws IOException {
+        super(name, factories, context, parent, pipelineAggregators, metaData);
+        this.separator = separator;
+        this.keys = keys;
+        this.filters = filters;
+        this.totalNumIntersections = ((keys.length * keys.length) - keys.length) / 2;
+        this.totalNumKeys = keys.length + totalNumIntersections;
+    }
+
+    private static class BitsIntersector implements Bits {
+        Bits a;
+        Bits b;
+
+        public BitsIntersector(Bits a, Bits b) {
+            super();
+            this.a = a;
+            this.b = b;
+        }
+
+        @Override
+        public boolean get(int index) {
+            return a.get(index) && b.get(index);
+        }
+
+        @Override
+        public int length() {
+            return Math.max(a.length(), b.length());
```

Comment:
It will be the same length in both cases, but I think `min` would be more correct for an intersection?